ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh

```diff
@@ -379,6 +414,148 @@ function upgrade() {
+        #       initial -> target -> ${SHARED_DIR}/upgrade-edge
+        #  if the minor version in ${SHARED_DIR}/upgrade-edge equal to target, then
+        #       1. we find a target+1 version as third target upgrade version
+        #       2. initial -> ${SHARED_DIR}/upgrade-edge -> target -> third target version
+        #  this case not allow ${SHARED_DIR}/upgrade-edge less than target
+
+        # INTERMEDIATE_MINOR_VERSIONT is the minor version in ${SHARED_DIR}/upgrade-edge
+        local retry=0 intermediate_image INTERMEDIATE_MINOR_VERSIONT block_second first_upgrade_to_image second_upgrade_to_image third_upgrade_to_image output latest_minor_version
+        intermediate_image="$(< "${SHARED_DIR}/upgrade-edge")"
+        if [[ -z "${intermediate_image:-}" ]]; then
+            echo "Error: The intermediate image is not given, break the job"
+            return 1
+        fi
+        
+        echo "Prepare test data"
+        INTERMEDIATE_MINOR_VERSIONT="$(oc adm release info -ojson "$intermediate_image" | jq -r '.metadata.version' | cut -f2 -d.)"
+        if [[ "$INTERMEDIATE_MINOR_VERSIONT" -gt "$TARGET_MINOR_VERSION" ]]; then
+            # This is y stream retarget upgrade, e.g.: 4.y -> (4.y | 4.y+1) -> (4.y+1 | 4.y+2)
+            first_upgrade_to_image=${TARGET}
+            second_upgrade_to_image=${intermediate_image}
+            block_second=true
+        else
+            if [[ "$INTERMEDIATE_MINOR_VERSIONT" == "$TARGET_MINOR_VERSION" ]]; then
+                # This is z stream retarget upgrade, e.g.: 4.y -> (4.y | 4.y+1) -> (4.y | 4.y+1)
+                first_upgrade_to_image=${intermediate_image}
+                second_upgrade_to_image=${TARGET}
+                
+                if oc adm release info -o jsonpath='{.digest}' quay.io/openshift-release-dev/ocp-release:4.$((TARGET_MINOR_VERSION+1)).0-ec.0-x86_64 2>&1; then
+                    # if third_upgrade_to_image not empty, we will retarget to it, but it will always be blocked
+                    # this can make sure we always have z and y stream retarget in one job
+                    latest_minor_version=$(( TARGET_MINOR_VERSION+1 ))
+                    third_upgrade_to_image="$( oc adm release info -o jsonpath='{.digest}' quay.io/openshift-release-dev/ocp-release:4.${latest_minor_version}.0-ec.0-x86_64 )"
+                fi
+
+                block_second=false
+            else
+                # If INTERMEDIATE_MINOR_VERSIONT < TARGET_MINOR_VERSION, then we will naver be able to upgrade to TARGET_MINOR_VERSION
+                # So do not put a small version in upgrade-edge
+                echo "Error: OCP-25473 do not cover rollback, break the job"
+                return 1
+            fi
+        fi
+        # first_upgrade_to_image can be both nightly build or stable build
+        if ! check_signed "${first_upgrade_to_image}"; then
+            FORCE_UPDATE="true"
+        else
+            FORCE_UPDATE="false"
+            admin_ack
```

Comment:
When FORCE_UPDATE="false", `cco_annotation` is needed before upgrade?

Reply:
done